mmdet/models/detectors/deformable_detr.py

```diff
@@ -1,12 +1,773 @@
+            grid = torch.cat([grid_x.unsqueeze(-1), grid_y.unsqueeze(-1)], -1)
+
+            scale = torch.cat([valid_W, valid_H], 1).view(num_feat, 1, 1, 2)
+            grid = (grid.unsqueeze(0).expand(num_feat, -1, -1, -1) +
+                    0.5) / scale
+            wh = torch.ones_like(grid) * 0.05 * (2.0**lvl)
+            proposal = torch.cat((grid, wh), -1).view(num_feat, -1, 4)
+            proposals.append(proposal)
+            _cur += (H * W)
+        output_proposals = torch.cat(proposals, 1)
+        output_proposals_valid = ((output_proposals > 0.01) &
+                                  (output_proposals < 0.99)).all(
+                                      -1, keepdim=True)
+        # inverse_sigmoid
+        output_proposals = torch.log(output_proposals / (1 - output_proposals))
+        output_proposals = output_proposals.masked_fill(
+            memory_mask.unsqueeze(-1), float('inf'))
+        output_proposals = output_proposals.masked_fill(
+            ~output_proposals_valid, float('inf'))
+
+        output_memory = memory
+        output_memory = output_memory.masked_fill(
+            memory_mask.unsqueeze(-1), float(0))
+        output_memory = output_memory.masked_fill(~output_proposals_valid,
+                                                  float(0))
+        output_memory = self.memory_trans_fc(output_memory)
+        output_memory = self.memory_trans_norm(output_memory)
+        # [bs, sum(hw), 2]
+        return output_memory, output_proposals
+
+    @staticmethod
+    def get_proposal_pos_embed(proposals: Tensor,
+                               num_pos_feats: int = 128,
+                               temperature: int = 10000):
+        """Get the position embedding of the proposal.
+
+        Args:
+            proposals (Tensor): Not normalized proposals, has shape
+                (bs, num_query, 4).
+            num_pos_feats (int, optional): The feature dimension for each
+                position along x, y, w, and h-axis. Note the final returned
+                dimension for each position is 4 times of num_pos_feats.
+                Default to 128.
+            temperature (int, optional): The temperature used for scaling the
+                position embedding. Defaults to 10000.
+
+        Returns:
+            Tensor: The position embedding of proposal, has shape
+            (bs, num_query, num_pos_feats * 4)
+        """
+        scale = 2 * math.pi
+        dim_t = torch.arange(
+            num_pos_feats, dtype=torch.float32, device=proposals.device)
+        dim_t = temperature**(2 * (dim_t // 2) / num_pos_feats)
+        # N, L, 4
+        proposals = proposals.sigmoid() * scale
+        # N, L, 4, 128
+        pos = proposals[:, :, :, None] / dim_t
+        # N, L, 4, 64, 2
+        pos = torch.stack((pos[:, :, :, 0::2].sin(), pos[:, :, :, 1::2].cos()),
+                          dim=4).flatten(2)
+        return pos
+
+
+class DeformableDetrTransformerEncoder(DetrTransformerEncoder):
```

Comment:
For `DeformableDetrTransformerEncoder`, `DeformableDetrTransformerDecoder`, 'DetrTransformerEncoderLayer' and 'DetrTransformerDecoderLayer',
Why not put them into `layers/` fold like `DetrTransformerEncoder`?
Why not import them as modules written into `__init__.py`?

Reply:
@JingweiZhang12 Hi~ Thanks for your review!~

>  Why not put them into layers/ fold like DetrTransformerEncoder?

In fact, the `XXXTransformerEncoder` and `XXXTransformerDecoder` are in layers/transformer.py originally.

We thought it makes `transformer.py` too long and decided to move them out.
We thought about making new files in `layers` like `detection_transformer_encoders.py` and `detection_transformer_decoders.py`. But I think it's more convenient for users in detector file.

_Reason 1:_ In this manner, in most cases, the code jumps do not require out of the file, or are close to the current file.

_Reason 2:_ In this manner, the code look very intuitive. If all the encodes and decoders are put together, the user may have a strong sense of fragmentation when reading the code.

_Reason 3:_ As the DETR-series algorithms develop, the length of this file will continue to increase, which may lead to some trouble.

_Reason 4:_
The original implementation build the encoders and the decoders with the 'registry' mechanism.
They are designed to be treated as compatible components of `Transformer` of the `bbox_head`.

In the refactor implementation, we treat the `detector` as the `Transformer`, so most logic in the detector modules is about `Transformer`. In this scheme, the `XXXTransformerEncoder` and `XXXTransformerDecoder` are designed to be treated as customized parts of each Transformer detector. So I prefer to keep these modules together for relevance, rather than splitting them up simply for compartmentalization.

>  Why not import them as modules written into __init__.py?

In fact it can, we just haven't understood the need to write them into `__init__.py`.
We'll discuss this !~


Looking forward to your feedback !~

Reply:
In some downstream codebases, we may only want to use and import `DeformableDetrTransformerEncoder` or `DeformableDetrTransformerEncoderLayer` from mmdetection. Thus,

`from mmdet.models import DeformableDetrTransformerEncoder` may be better than `from mmdet.modes.detectors.deformable_detr import DeformableDetrTransformerEncoder`.

Given the above,  it's better to put these modules into `layers/` rather `detector/` at the semantic level.

Reply:
@JingweiZhang12 Thanks for your suggention! The information is important for us! We'll discuss this!

Reply:
@JingweiZhang12 Hi, we agree with you. And here is a scheme:

Make a new folder in layers: /mmdet/models/layers/transformers/
and make a separate file for encoder and decoder of each DETR.

```text
/mmdet/models/layers
    /transformers
        detr_transformer.py
            class DetrTransformerEncoder
            class DetrTransformerDecoder
        deformable_detr_transformer.py
            class DeformableDetrTransformerEncoder
            class DeformableDetrTransformerDecoder
        conditional_detr_transformer.py
        dab_detr_transformer.py
        dino_transformer.py

    transformer.py
        class PatchEmbed
        class PatchMerging
        class DynamicConv
```

I would like to ask for your opinion again before revising.
Thanks for your helps! Look forward to your reply.